File: src/main/java/pw/mihou/nexus/features/command/core/NexusCommandEventCore.java
package pw.mihou.nexus.features.command.core;

import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import pw.mihou.nexus.features.command.facade.NexusCommand;
import pw.mihou.nexus.features.command.facade.NexusCommandEvent;

import java.util.HashMap;
import java.util.Map;

public class NexusCommandEventCore implements NexusCommandEvent {

    private final SlashCommandCreateEvent event;
    private final NexusCommand command;
    private final Map<String, Object> store = new HashMap<>();

    /**
     * Creates a new Nexus Event Core that is sent along with the Command Interceptors
     * and other sorts of handlers.
     *
     * @param event The base event received from Javacord.
     * @param command The command instance that is used.
     */
    public NexusCommandEventCore(SlashCommandCreateEvent event, NexusCommand command) {
        this.event = event;
        this.command = command;
    }

    @Override
    public SlashCommandCreateEvent getBaseEvent() {
        return event;
    }

    @Override
    public NexusCommand getCommand() {
        return command;
    }

    @Override
    public Map<String, Object> store() {
        return store;
    }
}

File: src/main/java/pw/mihou/nexus/features/command/core/NexusCommandEventCore.kt
package pw.mihou.nexus.features.command.core

import org.javacord.api.event.interaction.SlashCommandCreateEvent
import org.javacord.api.interaction.callback.InteractionOriginalResponseUpdater
import pw.mihou.nexus.features.command.facade.NexusCommand
import pw.mihou.nexus.features.command.facade.NexusCommandEvent
import pw.mihou.nexus.features.command.responses.NexusAutoResponse
import pw.mihou.nexus.features.commons.Deferrable
import pw.mihou.nexus.features.messages.NexusMessage
import java.util.concurrent.CompletableFuture
import java.util.concurrent.atomic.AtomicReference
import java.util.function.Function

class NexusCommandEventCore(override val event: SlashCommandCreateEvent, override val command: NexusCommand) : NexusCommandEvent {
    private val store: MutableMap<String, Any> = HashMap()
    var updater: AtomicReference<CompletableFuture<InteractionOriginalResponseUpdater>?> = AtomicReference(null)

    override fun store(): MutableMap<String, Any> = store

    override fun autoDefer(ephemeral: Boolean, response: Function<Void?, NexusMessage>): CompletableFuture<NexusAutoResponse> =
        Deferrable.autoDefer(event.slashCommandInteraction, updater, ephemeral, response)

    override fun respondLater(): CompletableFuture<InteractionOriginalResponseUpdater> {
        return updater.updateAndGet { interaction.respondLater() }!!
    }

    override fun respondLaterEphemerally(): CompletableFuture<InteractionOriginalResponseUpdater> {
        return updater.updateAndGet { interaction.respondLater(true) }!!
    }
}